File: app/src/main/java/com/tencent/iotvideo/link/popup/DeviceSettingDialog.java
package com.tencent.iotvideo.link.popup;

import android.content.Context;
import android.content.SharedPreferences;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.tencent.iot.twcall.R;
import com.tencent.iotvideo.link.util.VoipSetting;

import static android.content.Context.MODE_PRIVATE;

public class DeviceSettingDialog extends IosCenterStyleDialog {

    private EditText mJsonCopyEt;
    private EditText mProductEt;
    private EditText mDeviceNameEt;
    private EditText mDeviceKeyEt;
    private Button mConfirmBtn;

    public DeviceSettingDialog(Context context) {
        super(context, R.layout.popup_device_setting_layout);
    }

    @Override
    public void initView() {
        mJsonCopyEt = view.findViewById(R.id.et_json_copy);
        mProductEt = view.findViewById(R.id.et_login_product_id);
        mDeviceNameEt = view.findViewById(R.id.et_login_device_name);
        mDeviceKeyEt = view.findViewById(R.id.et_login_device_key);
        mConfirmBtn = view.findViewById(R.id.btn_confirm);

        mProductEt.setText(VoipSetting.getInstance(getContext()).productId);
        mDeviceNameEt.setText(VoipSetting.getInstance(getContext()).deviceName);
        mDeviceKeyEt.setText(VoipSetting.getInstance(getContext()).deviceKey);

        mConfirmBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (!checkDeviceInfo()) return;
                saveDeviceInfo();
                dismiss();
            }
        });

        mJsonCopyEt.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence s, int i, int i1, int i2) {
                String inputText = s.toString();
                if (!TextUtils.isEmpty(inputText)) {
                    if (VoipSetting.isJSONString(inputText)) {
                        VoipSetting.getInstance(getContext()).saveData(inputText);
                        VoipSetting.getInstance(getContext()).loadValueToMemory();
                        mProductEt.setText(VoipSetting.getInstance(getContext()).productId);
                        mDeviceNameEt.setText(VoipSetting.getInstance(getContext()).deviceName);
                        mDeviceKeyEt.setText(VoipSetting.getInstance(getContext()).deviceKey);
                    } else {
                        Toast.makeText(getContext(), "输入的json非法！", Toast.LENGTH_LONG).show();
                    }
                }
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });
    }

    private boolean checkDeviceInfo() {
        String productId = mProductEt.getText().toString();
        String deviceName = mDeviceNameEt.getText().toString();
        String deviceKey = mDeviceKeyEt.getText().toString();
        if (productId.isEmpty() || deviceName.isEmpty() || deviceKey.isEmpty()) {
            Toast.makeText(getContext(), "请输入设备信息！", Toast.LENGTH_LONG).show();
            return false;
        }
        return true;
    }

    private void saveDeviceInfo() {
        VoipSetting.getInstance(getContext()).setProductId(mProductEt.getText().toString());
        VoipSetting.getInstance(getContext()).setDeviceName(mDeviceNameEt.getText().toString());
        VoipSetting.getInstance(getContext()).setDeviceKey(mDeviceKeyEt.getText().toString());
    }
}

File: app/src/main/java/com/example/ivdemo/popup/DeviceSettingDialog.kt
package com.example.ivdemo.popup

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.RadioButton
import android.widget.Toast
import androidx.fragment.app.FragmentManager
import com.tencent.iot.twcall.databinding.PopupDeviceSettingLayoutBinding
import com.tencent.iotvideo.link.util.DeviceSetting
import com.tencent.iotvideo.link.util.updateOperate

class DeviceSettingDialog(private val context: Context) :
    IosCenterStyleDialog<PopupDeviceSettingLayoutBinding>(context) {

    private val deviceSetting by lazy { DeviceSetting.getInstance() }
    private var ipcType: Int = 2
    private val textWatcher = object : TextWatcher {
        override fun beforeTextChanged(
            s: CharSequence?,
            start: Int,
            count: Int,
            after: Int
        ) {
        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            val isOperate = checkDeviceInfo()
            binding.btnConfirm.updateOperate(isOperate)
        }

        override fun afterTextChanged(s: Editable?) {
        }

    }

    override fun getViewBinding(): PopupDeviceSettingLayoutBinding =
        PopupDeviceSettingLayoutBinding.inflate(layoutInflater)

    public override fun initView() {
        with(binding) {
            etLoginProductId.setText(deviceSetting.productId)
            etLoginDeviceName.setText(deviceSetting.deviceName)
            etLoginDeviceKey.setText(deviceSetting.deviceKey)
            val isOperate =
                !(deviceSetting.productId.isEmpty() || deviceSetting.deviceName.isEmpty() || deviceSetting.deviceKey.isEmpty())
            binding.btnConfirm.updateOperate(isOperate)

            etLoginProductId.addTextChangedListener(textWatcher)
            etLoginDeviceName.addTextChangedListener(textWatcher)
            etLoginDeviceKey.addTextChangedListener(textWatcher)
            ipcType = deviceSetting.ipcType
            rbOneWay.isChecked = ipcType == 1
            rbTwoWay.isChecked = ipcType == 2
            rgSelectWay.setOnCheckedChangeListener { group, checkedId ->
                ipcType = group.findViewById<RadioButton>(checkedId).tag.toString().toInt()
            }

            btnConfirm.setOnClickListener(View.OnClickListener {
                if (!checkDeviceInfo()) {
                    Toast.makeText(context.applicationContext, "请输入设备信息！", Toast.LENGTH_SHORT).show()
                    return@OnClickListener
                }
                saveDeviceInfo()
                dismiss()
                onDismiss?.invoke()
            })
        }
    }

    private fun checkDeviceInfo(): Boolean {
        val productId = binding.etLoginProductId.text.toString()
        val deviceName = binding.etLoginDeviceName.text.toString()
        val deviceKey = binding.etLoginDeviceKey.text.toString()
        return !(productId.isEmpty() || deviceName.isEmpty() || deviceKey.isEmpty())
    }

    private fun saveDeviceInfo() {
        with(binding) {
            deviceSetting.ipcType = ipcType
            deviceSetting.productId = etLoginProductId.text.toString()
            deviceSetting.deviceName = etLoginDeviceName.text.toString()
            deviceSetting.deviceKey = etLoginDeviceKey.text.toString()
        }
    }

    fun show(manager: FragmentManager) {
        super.show(manager, "DeviceSettingDialog")
    }

    private var onDismiss: (() -> Unit)? = null
    fun setDismissListener(onDismiss: () -> Unit) {
        this.onDismiss = onDismiss
    }
}
